java-extensions/jni-connector/src/main/java/com/starrocks/jni/connector/OffHeapColumnVector.java

```diff
@@ -0,0 +1,341 @@
+        } else if (type == OffHeapColumnType.STRING || type == OffHeapColumnType.DATE) {
+            int childCapacity = capacity;
+            childCapacity *= DEFAULT_ARRAY_LENGTH;
+            this.childColumns = new OffHeapColumnVector[1];
+            this.childColumns[0] = new OffHeapColumnVector(childCapacity, OffHeapColumnType.BYTE);
+            this.lengthData =
+                    Platform.reallocateMemory(lengthData, oldCapacity * 4L, newCapacity * 4L);
+            this.offsetData =
+                    Platform.reallocateMemory(offsetData, oldCapacity * 4L, newCapacity * 4L);
+        } else {
+            throw new RuntimeException("Unhandled " + type);
+        }
+        this.nulls = Platform.reallocateMemory(nulls, oldCapacity, newCapacity);
+        Platform.setMemory(nulls + oldCapacity, (byte) 0, newCapacity - oldCapacity);
+        capacity = newCapacity;
+    }
+
+    private void reset() {
+        if (childColumns != null) {
+            for (OffHeapColumnVector c : childColumns) {
+                c.reset();
+            }
+        }
+        elementsAppended = 0;
+        if (numNulls > 0) {
+            putNotNulls(0, capacity);
+            numNulls = 0;
+        }
+    }
+
+    private OffHeapColumnVector arrayData() {
+        return childColumns[0];
+    }
+
+    public boolean isNullAt(int rowId) {
+        return Platform.getByte(null, nulls + rowId) == 1;
+    }
+
+    public boolean hasNull() {
+        return numNulls > 0;
+    }
+
+    private void putNotNulls(int rowId, int count) {
+        if (!hasNull()) {
+            return;
+        }
+        long offset = nulls + rowId;
+        for (int i = 0; i < count; ++i, ++offset) {
+            Platform.putByte(null, offset, (byte) 0);
+        }
+    }
+
+    public int appendNull() {
+        reserve(elementsAppended + 1);
+        putNull(elementsAppended);
+        return elementsAppended++;
+    }
+
+    private void putNull(int rowId) {
+        Platform.putByte(null, nulls + rowId, (byte) 1);
+        ++numNulls;
+    }
+
+    public int appendBoolean(boolean v) {
+        reserve(elementsAppended + 1);
+        putBoolean(elementsAppended, v);
+        return elementsAppended++;
+    }
+
+    private void putBoolean(int rowId, boolean value) {
+        Platform.putByte(null, data + rowId, (byte) ((value) ? 1 : 0));
+    }
+
+    public boolean getBoolean(int rowId) {
+        return Platform.getByte(null, data + rowId) == 1;
+    }
+
+    public int appendShort(short v) {
+        reserve(elementsAppended + 1);
+        putShort(elementsAppended, v);
+        return elementsAppended++;
+    }
+
+    private void putShort(int rowId, short value) {
+        Platform.putShort(null, data + 2L * rowId, value);
+    }
+
+    public short getShort(int rowId) {
+        return Platform.getShort(null, data + 2L * rowId);
+    }
+
+    public int appendInt(int v) {
+        reserve(elementsAppended + 1);
```

Comment:
Is it possiable add a heap buffer ?

Reply:
could you provide more details?

Reply:
Writing directly to memory via the JVM may not be the most efficient. If append_int is called frequently we may generate a lot of function calls.
It is recommended to write to bytebuffer first, and then flush to C++ memory via memorycopy
eg:
```
appendInt(int a) {
reserve(size() + 1);
data[i++] = a;
}

flush() {
 memcpy(address, a, size())
}

```